lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -2781,12 +2783,329 @@ ModulePass *llvm::createDxilAllocateResourcesForLibPass() {
+// Helper class for legalizing dynamic resource use
+// Convert select/phi on resources to select/phi on index.
+// TODO: support case when save dynamic resource as local array element.
+// TODO: share code with LegalizeResourceUseHelper.
+class LegalizeDynamicResourceUseHelper {
+
+public:
+  ResourceUseErrors m_Errors;
+  ValueToValueMap HandleSelToIdxSel;
+  // Value sets we can use to iterate
+  ValueSetVector HandleSelects;
+
+  void CreateSelectsForHandleSelects() {
+    if (HandleSelects.empty())
+      return;
+
+    LLVMContext &Ctx = HandleSelects[0]->getContext();
+    Type *i32Ty = Type::getInt32Ty(Ctx);
+    Value *UndefValue = UndefValue::get(i32Ty);
+    // Create select for each HandleSelects.
+    for (auto &Select : HandleSelects) {
+      if (PHINode *Phi = dyn_cast<PHINode>(Select)) {
+        IRBuilder<> B(Phi);
+        unsigned numIncoming = Phi->getNumIncomingValues();
+        PHINode *newPhi = B.CreatePHI(i32Ty, numIncoming);
+        for (unsigned j = 0; j < numIncoming; j++) {
+          // Set incoming values to undef until next pass
+          newPhi->addIncoming(UndefValue, Phi->getIncomingBlock(j));
+        }
+        HandleSelToIdxSel[Phi] = newPhi;
+      } else if (SelectInst *Sel = dyn_cast<SelectInst>(Select)) {
+        IRBuilder<> B(Sel);
+        Value *newSel =
+            B.CreateSelect(Sel->getCondition(), UndefValue, UndefValue);
+        HandleSelToIdxSel[Sel] = newSel;
+      } else {
+        DXASSERT(false, "otherwise, non-select/phi in Selects set");
+      }
+    }
+  }
+
+  Value *getDynamicResourceIndex(Value *Hdl) {
+    if (isa<PHINode>(Hdl) || isa<SelectInst>(Hdl)) {
+      return HandleSelToIdxSel[Hdl];
+    } else if (CallInst *CI = dyn_cast<CallInst>(Hdl)) {
+      if (!hlsl::OP::IsDxilOpFuncCallInst(CI))
+        return nullptr;
+      DxilInst_CreateHandleFromHeap createHdl(CI);
+      if (createHdl)
+        return CI;
+      DxilInst_AnnotateHandle annotHdl(CI);
+      if (annotHdl)
+        return getDynamicResourceIndex(annotHdl.get_res());
+      return nullptr;
+    } else {
+      DXASSERT(false, "otherwise, non-select/phi in HandleSelects set");
+      return nullptr;
+    }
+  }
+
+  Value *getIsSampler(Value *Hdl) {
+    if (PHINode *Phi = dyn_cast<PHINode>(Hdl)) {
+      for (Value *V : Phi->incoming_values()) {
+        if (isa<PHINode>(V))
+          continue;
+        return getIsSampler(V);
+      }
+
+      DXASSERT(false, "fail to find createHandleFromHeap");
+      return nullptr;
+    } else if (SelectInst *Sel = dyn_cast<SelectInst>(Hdl)) {
+      return getIsSampler(Sel->getTrueValue());
+    } else if (CallInst *CI = dyn_cast<CallInst>(Hdl)) {
+      if (!hlsl::OP::IsDxilOpFuncCallInst(CI))
+        return nullptr;
+      DxilInst_CreateHandleFromHeap createHdl(CI);
+      if (createHdl)
+        return createHdl.get_samplerHeap();
+      DxilInst_AnnotateHandle annotHdl(CI);
+      if (annotHdl)
+        return getDynamicResourceIndex(annotHdl.get_res());
```

Comment:
Here is go check dynamic resource index for the input handle of annotateHandle.
But since we already sink annotateHandle after phi, we can just remove annotateHandle check here.